### PR TITLE
New dust plots

### PR DESCRIPTION
**Pull request by wmfw23**
Collection of new dust related plots:

- Dust-to-metal shaded phase diagram
- Hydrogen species fractions as a function of density, co-plotted with dust-to-metal ratio as a function of density
- Dust scaling relation plots fro data comparison collected from DeLooze et al (2020) data
- Dust to gas ratio as a function of oxygen abundance, to compare with Rémi-Ruyer et al (2014) data

NB. registration of some new quantities, change to VR file to add reading of mass-weighted Oxygen abundances. species fraction plot uses interpolated Ploeckinger et al (2020) tables for dust comparison. Added config arg points to an instance of this, but if unavailable will not plot. Also want to investigate selection function for DeLooze plots

### CODE REVIEW

**Comment by JBorrow** [comment]
Hi, looks like this was created against an older version of the repo. Can you fix the merge conflicts before I review?

File: colibre/config.yml [hunk 1]
```diff
@@ -10,7 +10,7 @@ auto_plotter_registration: registration.py
 
 # Location of the 'observational data' repository and its compiled
 # contents.
-observational_data_directory: ../observational_data
+observational_data_directory: ../../observational_data
```
Comment (JBorrow) [inline colibre/config.yml:13]: Why this change? Was this broken before?

File: colibre/config.yml [hunk 2]
```diff
@@ -41,6 +41,13 @@ scripts:
     output_file: density_temperature_dust_to_metals.png
     section: Density-Temperature
     title: Density-Temperature (Dust / Metals)
+  - filename: scripts/density_temperature_dust2metal.py
+    caption: Density-temperature diagram shaded by the total fraction of metals in the dust phase.
+    output_file: density_temperature_dust2metal.png
+    section: Density-Temperature
+    title: Dust-to-metal Ratio
+    additional_arguments:
+      quantity_type: hydro
```
Comment (JBorrow) [inline colibre/config.yml:50]: This appears to be repeated below

File: colibre/config.yml [hunk 3]
```diff
@@ -156,27 +163,42 @@ scripts:
...
+    caption: Co-plot of species fractions (left y-axis) and the dust-to-metal ratio (right y-axis) as a function opf gas density. If available, solid green line shows explicitly modelled dust-to-metal ratio, while dashed green is interpolated from the Ploeckinger+20 tables.
+    output_file: density_vs_species_fraction.png
+    section: Hydrogen Phases
+    title: Hydrogen Phase Fractions
+    additional_arguments:
+      cooling_tables: /cosma7/data/dp004/wmfw23/colibre_dust/coolingtables/UV_dust1_CR1_G1_shield1.hdf5
+      quantity_type: hydro
   - filename: scripts/density_temperature_species.py
-    caption: Density-temperature diagram shaded by H$_2$ mass fraction. The fraction is computed as the H$_2$ mass contained in each cell over the mass of gas in that cell.
+    caption: Density-temperature diagram shaded by H2 mass fraction.
```
Comment (JBorrow) [inline colibre/config.yml:182]: Can we keep the latex-ing in these, please? So H$_2$. I was also requested to add that additional comment, so if it's incorrect can we replace it with a correct alternative?

File: colibre/config.yml [hunk 4]
```diff
@@ -156,27 +163,42 @@ scripts:
...
-    section: Hydrogen Phase Density-Temperature
-    title: H$_2$
+    section: Hydrogen Phases
+    title: H$_2$ subgrid phase diagram
     additional_arguments:
       hydrogen_species: H2
       quantity_type: subgrid
   - filename: scripts/density_temperature_species.py
-    caption: Density-temperature diagram shaded by HI mass fraction. The fraction is computed as the HI mass contained in each cell over the mass of gas in that cell.
+    caption: Density-temperature diagram shaded by HI mass fraction.
```
Comment (JBorrow) [inline colibre/config.yml:190]: Similar comment above about how the binning is done.

File: colibre/config.yml [hunk 5]
```diff
@@ -156,27 +163,42 @@ scripts:
...
-    section: Hydrogen Phase Density-Temperature
-    title: HI
+    section: Hydrogen Phases
+    title: HI subgrid phase diagram
     additional_arguments:
       hydrogen_species: HI
       quantity_type: subgrid
   - filename: scripts/density_temperature_species.py
-    caption: Density-temperature diagram shaded by HII mass fraction. The fraction is computed as the HII mass contained in each cell over the mass of gas in that cell.
+    caption: Density-temperature diagram shaded by HII mass fraction.
```
Comment (JBorrow) [inline colibre/config.yml:198]: Same again here.

File: colibre/registration.py [hunk 6]
```diff
@@ -111,6 +111,8 @@
     except AttributeError:
         pass
 
+metal_frac = catalogue.apertures.zmet_gas_sf_100_kpc
+nonmetal_frac = 1.-metal_frac
```
Comment (JBorrow) [inline colibre/registration.py:115]: File needs formatting with black

File: colibre/registration.py [hunk 7]
```diff
@@ -143,9 +145,46 @@
 total_dust_mass = total_dust_fraction * catalogue.masses.m_gas
 name = f"$M_{{\\rm dust}}${dust_frac_error}"
 total_dust_mass.name = name
-nonmetal_frac = 1. - catalogue.apertures.zmet_gas_sf_100_kpc
+total_dust_fraction.name = f"$\\mathcal{{DTG}}${dust_frac_error}"
+total_dust_mass = total_dust_fraction * catalogue.masses.m_gas
```
Comment (JBorrow) [inline colibre/registration.py:149]: Should we be using the gas mass in the 100kpc aperture here?

File: colibre/registration.py [hunk 8]
```diff
@@ -143,9 +145,46 @@
...
+# Get depletion properties
+try:
+    gas_mass = catalogue.masses.m_gas
+    H_frac = getattr(catalogue.element_mass_fractions, "element_0")
+    O_frac = getattr(catalogue.element_mass_fractions, "element_4")
+    
+    o_abundance = unyt.unyt_array(
+        12+np.log10(O_frac/(16*H_frac)), "dimensionless")
+    o_abundance.name = "$12+\\log_{10}({{\\rm O/H}})$" 
+    setattr(self, "gas_o_abundance", o_abundance)
```
Comment (JBorrow) [inline colibre/registration.py:172]: Can we include aperture information here?

File: colibre/registration.py [hunk 9]
```diff
@@ -143,9 +145,46 @@
...
+    
+except AttributeError:
+    # We did not produce these quantities.
+    setattr(
+        self,
+        "gas_o_abundance",
+        unyt.unyt_array(
+            catalogue.masses.m_gas**0,
```
Comment (JBorrow) [inline colibre/registration.py:180]: not sure that this has the correct units for the 'fallback' array.

File: colibre/registration.py [hunk 10]
```diff
@@ -175,6 +215,24 @@
...
+    dust_to_hi_ratio.name = "M_{{\\rm dust}}/M_{{\\rm HI}}"    
+    setattr(self, "gas_dust_to_hi_ratio", dust_to_hi_ratio)
+    
+except AttributeError:
+    # We did not produce these quantities.
+    setattr(
+        self,
+        "gas_dust_to_hi_ratio",
+        unyt.unyt_array(
+            catalogue.masses.m_gas**0,
```
Comment (JBorrow) [inline colibre/registration.py:230]: Same comments here as above, would be nice to have the units explicit.

File: colibre/registration.py [hunk 11]
```diff
@@ -175,6 +215,24 @@
         ),
     )
 
+# Get HI to dust ratio
+try:
+    dust_to_hi_ratio = total_dust_mass / HI_mass
+    dust_to_hi_ratio.name = "M_{{\\rm dust}}/M_{{\\rm HI}}"    
+    setattr(self, "gas_dust_to_hi_ratio", dust_to_hi_ratio)
```
Comment (JBorrow) [inline colibre/registration.py:222]: Aperture information? If not in an aperture, but uses FOF halo, would be nice to have a comment here explicitly stating that.

File: colibre/scripts/density_species_interp.py [hunk 12]
```diff
@@ -0,0 +1,339 @@
+"""
+Makes a rho-T plot. Uses the swiftsimio library.
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
+import glob
+
+import swiftsimio as sw
+print(sw.__file__)
```
Comment (JBorrow) [inline colibre/scripts/density_species_interp.py:10]: Please remove this print statement

File: colibre/scripts/density_species_interp.py [hunk 13]
```diff
@@ -0,0 +1,339 @@
+"""
+Makes a rho-T plot. Uses the swiftsimio library.
```
Comment (JBorrow) [inline colibre/scripts/density_species_interp.py:2]: Please add a longer decsription of the specific function of this rho-T instance.

File: colibre/scripts/density_species_interp.py [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+    "Carbon",
+    "Nitrogen",
+    "Oxygen",
+    "Neon",
+    "Magnesium",
+    "Silicon",
+    "Sulphur",
+    "Calcium",
+    "Iron",
+    "Oatoms",]
```
Comment (JBorrow) [inline colibre/scripts/density_species_interp.py:38]: Needs formatting with black

Comment (JBorrow) [inline colibre/scripts/density_species_interp.py:38]: Do we not have this information from the namedcolumns instance?

File: colibre/scripts/density_species_interp.py [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+    ionfrac = X * masses * data.gas.species_fractions.HII
+
+    # Dust Mass Fractions
+    dfracs = np.zeros(data.gas.masses.shape)
+    for d in dir(data.gas.dust_mass_fractions):
+        if hasattr(getattr(data.gas.dust_mass_fractions, d), "units"):
+            dfracs += getattr(data.gas.dust_mass_fractions, d)
+            
+    if glob.glob(tables):
+        table_file = h5.File(tables, "r")
```
Comment (JBorrow) [inline colibre/scripts/density_species_interp.py:69]: Please use `with h5.File(tables, "r")` rather than explitcitly opening and closing the file. That ensures if it crashes, or an exception is raised, that the file is closed 'nicely'

File: colibre/config.yml (hunk 1 above)
Comment (wmfw23) [reply colibre/config.yml]: my mistake, I use an external data directory that i can develop separately, but need to change this back on commit

File: colibre/config.yml (hunk 2 above)
Comment (wmfw23) [reply colibre/config.yml:50]: removed

File: colibre/config.yml (hunk 3 above)
Comment (wmfw23) [reply colibre/config.yml]: added these back, removed accidentally

File: colibre/config.yml (hunk 4 above)
Comment (wmfw23) [reply colibre/config.yml]: as above

File: colibre/config.yml (hunk 5 above)
Comment (wmfw23) [reply colibre/config.yml]: as above

File: colibre/registration.py (hunk 6 above)
Comment (wmfw23) [reply colibre/registration.py]: done

File: colibre/registration.py (hunk 7 above)
Comment (wmfw23) [reply colibre/registration.py:149]: use gas mass now

File: colibre/registration.py (hunk 8 above)
Comment (wmfw23) [reply colibre/registration.py:173]: This is a subhalo property, added comment as below

File: colibre/registration.py (hunk 9 above)
Comment (wmfw23) [reply colibre/registration.py]: the `**0` makes it dimensionless

File: colibre/scripts/density_species_interp.py (hunk 14 above)
Comment (wmfw23) [reply colibre/scripts/density_species_interp.py]: Unfortunately, there's an extra column in the cooling tables (`Oatoms`) I thought it was more explicit to write these out rather than use named columns and add one.

File: colibre/registration.py (hunk 10 above)
Comment (wmfw23) [reply colibre/registration.py]: it is the correct units, but put an explicit conversion in each one

File: colibre/registration.py (hunk 11 above)
Comment (wmfw23) [reply colibre/registration.py]: added comment as above

File: colibre/scripts/density_species_interp.py (hunk 12 above)
Comment (wmfw23) [reply colibre/scripts/density_species_interp.py]: done

File: colibre/scripts/density_species_interp.py (hunk 13 above)
Comment (wmfw23) [reply colibre/scripts/density_species_interp.py]: changed

File: colibre/scripts/density_species_interp.py (hunk 14 above)
Comment (wmfw23) [reply colibre/scripts/density_species_interp.py]: done

File: colibre/scripts/density_species_interp.py (hunk 15 above)
Comment (wmfw23) [reply colibre/scripts/density_species_interp.py]: done

**Comment by wmfw23** [comment]
overlooked this for a while, pushed new update

File: colibre/scripts/density_species_interp.py (hunk 14 above)
Comment (JBorrow) [reply colibre/scripts/density_species_interp.py]: I would prefer it if you used the named columns and added on "Oatoms", if at all possible.

Comment (wmfw23) [reply colibre/scripts/density_species_interp.py]: you also need to add Sulphur & Magnesium and remove Europium to get the same columns in an order dependent way - think it's better to write out
